src/alluvium/avro.py

```diff
@@ -0,0 +1,251 @@
+    ASCENDING = "ascending"
+    DESCENDING = "descending"
+    IGNORE = "ignore"
+
+
+class AvroSchema(ABC):
+    def __init__(self, avro_type: Union[AvroPrimitiveType, AvroComplexType]):
+        self.type = avro_type
+
+    @abstractmethod
+    def avro_schema(self):
+        pass
+
+
+class AvroRecordField(object):
+    def __init__(
+        self,
+        name: str,
+        avro_type: Union[AvroSchema, AvroPrimitiveType],
+        doc: str = None,
+        default: Any = None,
+        order: FieldOrdering = None,
+        aliases: List[str] = None,
+    ):
+        self.name = name
+        self.avro_type = avro_type
+        self.doc = doc
+        self.default = default
+        self.order = order
+        self.aliases = aliases
+
+    def generate_avro_field(self):
+        base = {
+            "type": resolve_primitive_type(self.avro_type)
+            if isinstance(self.avro_type, (AvroPrimitiveType, AvroCustomType))
+            else self.avro_type.avro_schema(),
+            "name": self.name,
+        }
+        if self.doc:
+            base["doc"] = self.doc
+        if self.default:
+            base["default"] = self.default
+        if self.order:
+            base["order"] = self.order.value
+        if self.aliases:
+            base["aliases"] = self.aliases
+        return base
+
+
+class AvroRecordSchema(AvroSchema):
+    avro_type = AvroComplexType.RECORD
+
+    def __init__(
+        self,
+        name: str,
+        fields: List[AvroRecordField],
+        namespace: str = None,
+        doc: str = None,
+        aliases: List[str] = None,
+    ):
+        if not fields:
+            raise InvalidSchemaException(
+                self.avro_type, "Cannot provide a record schema without any fields.", name=name
+            )
+        self.name = name
+        self.fields = fields
+        self.namespace = namespace
+        self.doc = doc
+        self.aliases = aliases
+
+    def avro_schema(self):
+        base = {
+            "type": self.avro_type.value,
+            "name": self.name,
+            "fields": [field.generate_avro_field() for field in self.fields],
+        }
+        if self.doc:
+            base["doc"] = self.doc
+        if self.namespace:
+            base["namespace"] = self.namespace
+        if self.aliases:
+            base["aliases"] = self.aliases
+        return base
+
+
+class AvroArraySchema(AvroSchema):
+    avro_type = AvroComplexType.ARRAY
+
+    def __init__(self, items: Union[AvroSchema, AvroPrimitiveType]):
+        self.items = items
+
+    def avro_schema(self):
+        return {
+            "type": self.avro_type.value,
+            "items": resolve_primitive_type(self.items)
+            if isinstance(self.items, (AvroPrimitiveType, AvroCustomType))
+            else self.items.avro_schema(),
+        }
+
+
+class AvroMapSchema(AvroSchema):
+    avro_type = AvroComplexType.MAP
+
+    def __init__(self, values: Union[AvroSchema, AvroPrimitiveType]):
```

Comment:
Doesn't a MAP type object need a key field along with a values field?

Reply:
According to the avro spec, the key for a map must be a string so it's redundant to include that info in the schema. I agree that that spec is super weird though : (